src/embed/connection.rs

```diff
@@ -0,0 +1,256 @@
+use super::dispatch::Request;
+use super::types::*;
+use crate::dag;
+use crate::db;
+use async_fn::AsyncFn2;
+use async_std::stream::StreamExt;
+use async_std::sync::{Receiver, RecvError, RwLock};
+use futures::stream::futures_unordered::FuturesUnordered;
+use log::warn;
+use nanoserde::{DeJson, SerJson};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU32, Ordering};
+
+lazy_static! {
+    static ref TRANSACTION_COUNTER: AtomicU32 = AtomicU32::new(1);
+}
+
+type TxnMap<'a> = RwLock<HashMap<u32, RwLock<db::Write<'a>>>>;
+
+fn deserialize<T: DeJson>(data: &str) -> Result<T, String> {
+    match DeJson::deserialize_json(data) {
+        Ok(v) => Ok(v),
+        Err(e) => Err(format!("InvalidJson({})", e)),
+    }
+}
+
+enum UnorderedResult {
+    Request(Result<Request, RecvError>),
+    Stop(),
+    None(),
+}
+
+async fn connection_future<'a, 'b>(
+    rx: &Receiver<Request>,
+    store: &'a dag::Store,
+    txns: &'b TxnMap<'a>,
+    request: Option<Request>,
+) -> UnorderedResult {
+    let req = match request {
+        None => return UnorderedResult::Request(rx.recv().await),
```

Comment:
I totally agree, but I think Rust's half-implemented async type system forbids this! I'd love to be proved wrong, however.

What I found when using FutureUnordered was that if I ever tried to push in two places with two different async functions, or even `async { ... }` blocks taking and returning nothing, Rust would say that I was trying to mix two different future types, I believe because it constructs a new and distinct fake type for each place `async` function or block, regardless of the types involved.

Making this single function that, when run, does different jobs based on its parameters was the only way I was able to work around this in my exploration.

CC @aboodman in case he has other ideas.

Reply:
>  I believe because it constructs a new and distinct fake type for each place async function or block, regardless of the types involved.

Correct.

Reply:
Here are some ideas which may or may not work:

1. You can use dynamic dispatch to construct a set of futures with a consistent type. The type param of the UnorderedSet would have to be something like `Pin<Box<dyn Future<...>>>`. You will have to either manually wrap the statically typed futures like I showed @phritz in a previous review or else maybe AsyncFn does this? I don't really recommend this path.

2. You can continue to have a single async function that runs each pump of the message loop but just decompose it into two sub functions and use better naming to try and make more clear what's going on. Or just use naming in this function as-is.

3. Stepping back, combining receive in this way with executing the functions is clever, but another way to factor it would be to have an UnorderedFutures instance only alive when something is executing, and have the read loop construct a new one as needed. I think that might be easier to read:

```
for msg in recv {
  if futures_unordered.is_complete() {
    futures_unordered = FuturesUnordered::new();
  }
  futures_unordered.push(connection_future(msg));
}
```

I think this might also mean you don't need an explicit stop message? I haven't thought it through completely.

Reply:
Ah, one detail that *may* prevent 3. is the documentation on FuturesUnordered::push: https://docs.rs/futures/0.3.5/futures/stream/futures_unordered/struct.FuturesUnordered.html#method.push. Basically, it sounds like we need to keep calling poll_next() on it or the newly-pushed futures wont be run, because these are being executed within this function's stack frame, rather than via `spawn` or something like that.

I think @phritz's `select!` idea is interesting, and that macro in Rust was unknown to me. I don't feel ambition to put my feet back into this fire though, so I'll add some comments describing why I did it this way and maybe improve some naming, but I'm not going to explore alternative implementation techniques at this time.

Reply:
Sounds good. Seems like all we need is some comments to orient the reader to what's going on so they don't have to work it out themselves.